src/drawing/polygon.rs

```diff
@@ -71,25 +66,29 @@ where
     y_min = max(0, min(y_min, height as i32 - 1));
     y_max = max(0, min(y_max, height as i32 - 1));
 
-    let mut closed = Vec::with_capacity(poly.len() + 1);
-    for p in poly {
-        closed.push(*p);
-    }
+    let mut closed: Vec<Point<i32>> = poly.iter().copied().collect();
     closed.push(poly[0]);
 
     let edges: Vec<&[Point<i32>]> = closed.windows(2).collect();
-    let mut intersections: Vec<i32> = Vec::new();
+    let mut intersections = Vec::new();
 
     for y in y_min..y_max + 1 {
         for edge in &edges {
             let p0 = edge[0];
             let p1 = edge[1];
 
             if p0.y <= y && p1.y >= y || p1.y <= y && p0.y >= y {
-                // Need to handle horizontal lines specially
                 if p0.y == p1.y {
+                    // Need to handle horizontal lines specially
                     intersections.push(p0.x);
                     intersections.push(p1.x);
+                } else if p0.y == y || p1.y == y {
+                    if p1.y > y {
+                        intersections.push(p0.x);
+                    }
+                    if p0.y > y {
+                        intersections.push(p1.x);
+                    }
```

Comment:
This is the main difference in the algorithm (handling the case when `y` is at the end of an edge).
There are 3 cases when calculating the intersection with `y` being in the end point of an edge:
- when both edges that end in `y` are above `y` (we will count them both)
![image](https://user-images.githubusercontent.com/40909265/95653608-242d9180-0afa-11eb-99a7-250e3907bc55.png)

- when both edges that end in `y` are bellow `y` (we will ignore them both)
![image](https://user-images.githubusercontent.com/40909265/95653615-2db6f980-0afa-11eb-8cb8-d0cd50aab6d1.png)

- when edges that end in `y` have their ends on the opposite sides compared to `y` (we will count just one)
![image](https://user-images.githubusercontent.com/40909265/95653617-34de0780-0afa-11eb-8d0e-b053c77f869d.png)

The previous algorithm was counting both ends always, which messed up the drawing in case of a concave polygon